### ISSUE TITLE
Submission: stochLAB

### ISSUE THREAD

**Issue by grwhumphries** [ropensci/software-review#551]
**Date accepted:** 2023-02-03

Submitting Author Name: Grant Humphries
Submitting Author Github Handle: @grwhumphries
Other Package Authors Github handles: (comma separated, delete if none) @bcaneco
Repository:  https://github.com/HiDef-Aerial-Surveying/stochLAB
Version submitted: 1.1.1
Submission type: Stats
Badge grade: gold
Editor: @stephaniehicks
Reviewers: @kstreet13, @clairemas0n
Due date for @kstreet13: 2022-11-01
Due date for @clairemas0n: 2022-11-24
Archive: TBD
Version accepted: TBD
Language: en
---

-   Paste the full DESCRIPTION file inside a code block below:

```
Type: Package
Package: stochLAB
Title: Stochastic Collision Risk Model
Version: 1.1.0
Authors@R: c(
    person("Grant", "Humphries", , "grant.humphries@hidefsurveying.co.uk", role = c("aut", "cre")),
    person("Bruno", "Caneco", , "bruno@dmpstats.com", role = "aut"),
    person("Marine Scotland", role = c("fnd", "cph")),
    person("HiDef", role = "cph"),
    person("DMPstats", role = "cph")
  )
Description: Tool for running Collision Risk Models for seabirds on
    offshore wind farms.
License: GPL (>= 3) + file LICENSE
Depends: 
    R (>= 4.0)
Imports: 
    cli,
    data.table,
    dplyr,
    foreach,
    glue,
    logr,
    magrittr,
    msm,
    pracma,
    purrr,
    rlang,
    stats,
    tibble,
    tidyr
Suggests: 
    rmarkdown,
    knitr,
    spelling,
    testthat (>= 3.0.0),
    covr
Config/testthat/edition: 3
Encoding: UTF-8
Language: en-GB
LazyData: true
LazyDataCompression: bzip2
Roxygen: list(markdown = TRUE)
URL: https://www.github.com/HiDef-Aerial-Surveying/stochLAB,
    https://hidef-aerial-surveying.github.io/stochLAB/
BugReports: https://www.github.com/HiDef-Aerial-Surveying/stochLAB/issues
RoxygenNote: 7.2.0
VignetteBuilder: knitr
```

## Pre-submission Inquiry

- [X] A pre-submission inquiry has been approved in [issue#506](506)

## General Information

-   Who is the target audience and what are scientific applications of this package?
This package is designed for consultants or scientists working in the offshore windfarm industry who are looking to quantify the risks of collision to seabirds

-   Paste your responses to our [*General Standard* **G1.1** here](https://stats-devguide.ropensci.org/standards.html#general-standards), describing whether your software is:

    - *The first implementation of a novel algorithm*; or
    - *The first implementation within* **R** *of an algorithm which has previously been implemented in other languages or contexts*; or
    - *An improvement on other implementations of similar algorithms in* **R**.
   
Core calculations follow the work developed by [Masden (2015)](https://data.marine.gov.scot/dataset/developing-avian-collision-risk-model-incorporate-variability-and-uncertainty)

    Please include hyperlinked references to all other relevant software.

-   (If applicable) Does your package comply with our [guidance around *Ethics, Data Privacy and Human Subjects Research*](https://devguide.ropensci.org/policies.html#ethics-data-privacy-and-human-subjects-research)?
- NA

## Badging

-    What grade of badge are you aiming for? ([bronze, silver, gold](https://stats-devguide.ropensci.org/pkgdev.html#pkgdev-badges))
Gold

-    If aiming for silver or gold, describe which of the [four aspects listed in the *Guide for Authors* chapter](https://stats-devguide.ropensci.org/pkgdev.html#pkgdev-silver) the package fulfils (at least one aspect for silver; three for gold)
Compliance with a good number of standards beyond those identified as minimally necessary. This will require reviewers and authors to agree on identification of both a minimal subset of necessary standards, and a full set of potentially applicable standards. This aspect may be considered fulfilled if at least one quarter of the additional potentially applicable standards have been met, and should definitely be considered fulfilled if more than one half have been met.

## Technical checks

Confirm each of the following by checking the box.

- [X] I have read the [rOpenSci packaging guide](https://devguide.ropensci.org/building.html).
- [X] I have read the [author guide](https://devdevguide.netlify.app/authors-guide.html) and I expect to maintain this package for at least 2 years or have another maintainer identified.
- [X] I/we have read the [*Statistical Software Peer Review* Guide for Authors](https://stats-devguide.ropensci.org/pkgdev.html).
- [X] I/we have run [`autotest`](https://github.com/ropensci-review-tools/autotest) checks on the package, and ensured no tests fail.
- [X] The [`srr_stats_pre_submit()` function](https://ropensci-review-tools.github.io/srr/reference/srr_stats_pre_submit.html) confirms this package may be submitted.
- [X] The [`pkgcheck()` function](https://docs.ropensci.org/pkgcheck/reference/pkgcheck.html) confirms this package may be submitted - alternatively, please explain reasons for any checks which your package is unable to pass.

This package:

- [X] does not violate the Terms of Service of any service it interacts with.
- [X] has a CRAN and OSI accepted license.
- [X] contains a [README with instructions for installing the development version](https://devguide.ropensci.org/building.html#readme).

## Publication options

- [X] Do you intend for this package to go on CRAN?
- [ ] Do you intend for this package to go on Bioconductor?

## Code of conduct

- [X] I agree to abide by [rOpenSci's Code of Conduct](https://ropensci.org/code-of-conduct/) during the review process and in maintaining my package should it be accepted.

**Comment by emilyriederer** [comment]
Hi @grwhumphries ! Thank you for making a full submission. We may experience a slight delay as @mpadge is offline next week and it appears we may need assistance with the editor check bot. 

In the meantime, could you please edit your message to link your to your presubmission inquiry in #506 for record-keeping?

Thanks!

**Comment by grwhumphries** [comment]
Hi @emilyriederer - thanks for that and no worries - just updated with the link to issue #506

**Comment by mpadge** [comment]
Thanks for the submission @grwhumphries, and for bearing with us while we use submissions to refine our system. In your case, the above "srr" section reveals a bug which we didn't anticipate. I can now appreciate that our [*Guide for Authors*](https://stats-devguide.ropensci.org/pkgdev.html) of statistical packages isn't clear enough, and needs a sentence at the outset along the lines of:

> All packages submitted for software peer review must document compliance with both General Standards, and at least one set of our category-specific standards.

Your package only documents compliance with the general standards, and needs to include statements of compliance with at least one specific category of standards. Updates to our checking system in response to your submission would now generate the following message:

---

### 1. rOpenSci Statistical Standards ([`srr` package](https://github.com/ropensci-review-tools/srr))



:heavy_multiplication_x: Error: Package documents compliance only with general standards. Statistical packages must document compliance with at least one set of category-specific standards as well.

---

I'll leave you to discuss with @emilyriederer whether you might need to put a "Holding" label on this submission while you address the missing standards. Please accept our apologies for any confusion.

**Comment by emilyriederer** [comment]
Thanks, @mpadge 

Hi @grwhumphries - per Mark's comment, please let us know what sort of timeline you prefer for addressing the standards. I can put a `holding` label in this issue in the meantime if you don't expect it to be a short-term priority. Whatever works best for you!

**Comment by grwhumphries** [comment]
Hi @emilyriederer and @mpadge  - sorry about the delay.   I think I can address this within the week - I just need to get my head back into it. Hope that's okay?

**Comment by emilyriederer** [comment]
Sounds great @grwhumphries ! No timeline pressure on our side. Only wanted to know about the tag for repo / issue-log hygiene but not a big deal either way.

**Comment by grwhumphries** [comment]
@mpadge and @emilyriederer  okay the latest version of `stochLAB` (v1.1.1) has just been pushed and I think that covers out the standards for EDA and PD.

**Comment by emilyriederer** [comment]
@ropensci-review-bot check package

**Comment by grwhumphries** [comment]
@emilyriederer @mpadge - I just realized that the package check failed because the PD standards aren't merged into the package check yet.  How do you want me to handle that? 

Cheers
G

**Comment by mpadge** [comment]
@emilyriederer One of the statistical categories this package complies with is "*Probability Distributions*", which is currently in [draft-mode only](https://github.com/ropensci/statistical-software-review-book/pull/67). I've implemented a corresponding "dev" branch of [the `srr` package](https://github.com/ropensci-review-tools/srr/pull/32) which the above checks use. These dev branches should be merged by mid-September 2022. Until then please note the following:

1. Automated checks will fail, because there is no way to trigger "dev" branch checks here, but we'll always see that fail and be able to ping the dev checks ourselves to dump results here.
2. Anybody wanting to check the package locally can easily do so by first running:
    ``` r
    remotes::install_github("ropensci-review-tools/srr@dev")
    ```
    All checks will then be re-directed to current "dev" branch of standards including the new categories;
3. We'll strive to get these new standards incorporated **before reviews of this package commence**. If we don't notify here, both @grwhumphries and reviewers will need to be warned that the *Probability Distributions Standards* may change during the review process, and so reviews *may* require iteration. (We will of course strive to avoid this!).

Thanks both, and great to have submissions really pushing our system along! :rocket: :bullettrain_side:

**Comment by grwhumphries** [comment]
@emilyriederer and @mpadge I've just re-run the package check with the srr dev branch and that seems to have passed. I also fixed the issue above (no returns documented) - which I guess is a new check as that wasn't picked up in previous checks?  It should be ready to go again now

**Comment by emilyriederer** [comment]
Thanks @grwhumphries ! This is all great news. We will begin looking for an editor and hope to connect you shortly

**Comment by emilyriederer** [comment]
Hi @grwhumphries - I'm delighted to introduce @stephaniehicks as the handling editor for this submission. Thank you both for your work on this!

**Comment by emilyriederer** [comment]
@ropensci-review-bot assign @stephaniehicks as editor

**Comment by grwhumphries** [comment]
@stephaniehicks - nice to meet you and looking forward to working with you

**Comment by stephaniehicks** [comment]
@ropensci-review-bot seeking reviewers

**Comment by stephaniehicks** [comment]
Hi @grwhumphries, great to meet you too. I apologize for my slow response on this. I'm working on identifying reviewers. 

@mpadge can you confirm that I have access to `@ropensci-review-bot seeking reviewers`. At the moment, it looks like I don't have access to it and will need you or another editor's help here to initiate the bot. Thanks!

**Comment by mpadge** [comment]
@stephaniehicks Sorry about that, an oversight on our part. You should now be able to access all bot commands.

**Comment by stephaniehicks** [comment]
@ropensci-review-bot seeking reviewers

**Comment by stephaniehicks** [comment]
@mpadge Thanks! It looks like it is working now. Appreciate the help.

**Comment by stephaniehicks** [comment]
@grwhumphries, feel free to add the badge to the README of your package 👆

**Comment by stephaniehicks** [comment]
@ropensci-review-bot add @kstreet13

**Comment by stephaniehicks** [comment]
@ropensci-review-bot assign @kstreet13 as reviewer

**Comment by stephaniehicks** [comment]
@mpadge just a small note about the stats dev guide: 

The Handling Editor Checklist (https://stats-devguide.ropensci.org/pkgsubmission.html#handling-editor-checklist) says to use `@ropensci-review-bot add <@GITHUB_USERNAME> to reviewers` to add a reviewer, but following `@ropensci-review-bot help`, I noticed it is `@ropensci-review-bot assign xxxxx as reviewer` (i.e. need to replace `add` with `assign`).

**Comment by stephaniehicks** [comment]
Hi everyone, I'd like to introduce Kelly Street (@kstreet13) (https://kellystreet.org) who is an Assistant Professor at UCS. He has kindly agreed to be a reviewer here. However, I wanted to mention that he will need some time to review the guide to go through the content there first. Just a heads up! Thank you again @kstreet13 and let me know if you have any questions!

**Comment by mpadge** [comment]
Sorry for confusion @stephaniehicks! The bot should recognise any one of "add to reviewers" "add as reviewer", "assign as reviewer", or "assign to reviewers". I think the [preceding glitch](https://github.com/ropensci/software-review/issues/551#issuecomment-1274757490) was because of missing "to reviewers". Happy that all now seems good :smile:

**Comment by stephaniehicks** [comment]
Ah! that makes sense. Thanks for the clarification!

**Comment by kstreet13** [comment]
Apologies for the delay, everybody. I was unclear on the expectations for reviewers and ended up wasting some effort checking various standards in great detail. I will have a review for you as soon as I can!

**Comment by stephaniehicks** [comment]
Thank you @kstreet13! Please let me know if there's anything I can do to help support you in the review process. We greatly appreciate your time and effort on this.

**Comment by stephaniehicks** [comment]
@ropensci-review-bot assign @clairemas0n as reviewer

**Comment by stephaniehicks** [comment]
Hi everyone, I'd like to welcome Claire Mason (@clairemas0n) to the thread! She has kindly agreed to be a reviewer here. Thank you again @clairemas0n and let me know if you have any questions!

**Comment by kstreet13** [comment]
Sorry again for the delay! As I said, this was my first time reviewing for ROpenSci and I was a little unclear on the expectations.

---

## Package Review

- [x] As the reviewer I confirm that there are no [conflicts of interest](https://devguide.ropensci.org/policies.html#coi) for me to review this work (If you are unsure whether you are in conflict, please speak to your editor _before_ starting your review).

---

### Compliance with Standards

- [x] This package complies with a sufficient number of standards for a silver badge
- [ ] This grade of badge is the same as what the authors wanted to achieve

The following standards currently deemed non-applicable (through tags of `@srrstatsNA`) could potentially be applied to future versions of this software:
 - **EA2.4** and **EA2.5** concern the use of multitabular input, which is relevant to both `stoch_crm` and `band_crm`.
 - **G5.2**, **G5.2a**, and **G5.2b**. The authors are correct that inputs are checked by the `validate_inputs` scripts, but this is not explicitly tested, as far as I can tell (ie. by having unit tests that give invalid inupts and checking for the correct error message).
 - **G1.5** is not directly applicable, but the authors do claim in the documentation that their implementation of the extended model is "over 30x faster than Masden's code", which is not demonstrated.
 - **PD3.5** seems to concern integration generally, which is a part of the model.


Please also comment on any standards which you consider either particularly well, or insufficiently, documented.
 - **G1.3** Details of the models including statistical terminology are not always explained, but users are directed to appropriate sources, such as Masden (2015) and Band (2012).
 - **G1.4a** Internal functions are not all documented (esp. in `validate_inputs`)


For packages aiming for silver or gold badges:

- [x] This package extends beyond minimal compliance with standards in the following ways:
Function documentation is quite extensive, which may be consequence of how complicated the core methods are, but is nonetheless impressive.

---

### General Review

#### Documentation

The package includes all the following forms of documentation:

- [ ] **A statement of need** clearly stating problems the software is designed to solve and its target audience in README
*The first sentence lays out what the package does, but I'm a little unclear on who the target audience is. I'm guessing it's some combination of people who own wind farms and ornithologists?*
- [x] **Installation instructions:** for the development version of package and any non-standard dependencies in README
- [x] **Community guidelines** including contribution guidelines in the README or CONTRIBUTING
- [x] The documentation is sufficient to enable general use of the package beyond one specific use case

#### Algorithms

- How well are algorithms encoded?
*The overall structure looks good to me, though some of the details (both of the method and the implementation) are lost on me. I'm guessing the big `for` loop that runs the model a bunch of times with different parameters is the most time-consuming part, so one potential future improvement would be to run those in parallel rather than sequentially.*

- Is the choice of computer language appropriate for that algorithm, and/or envisioned use of package?
*Definitely. This is a stats-heavy model, so R seems like a natural choice.*

- Are aspects of algorithmic scaling sufficiently documented and tested?
*I'm not sure this question is relevant. It (presumably) scales linearly with the number of iterations, but most functions parameters (that may or may not be inferred from data) rather than directly using any data.*

- Are there any aspects of algorithmic implementation which could be improved?
*The parallelization mentioned above could help in cases where users want to run more iterations or try more models. But overall, the examples run pretty quickly, so it seems to be pretty well optimized (and I'm not qualified to comment on the correctness of the implementation, but that seems to checked by the tests).*


#### Testing

- Regardless of actual coverage of tests, are there any fundamental software operations which are not sufficiently expressed in tests?
*I might be missing something, but I don't see any code coverage statistics. That said, tests seem to focus on assessing the correctness of the implementation. This is very important, but misses some of the more granular stuff like making sure every error/warning message works (as in **G5.8**). `test-validate_inputs.R` has some of this (and then some more, commented out), but given that there are SO MANY inputs to the main functions in this package, it definitely seems like these tests should be more extensive.*

- Is there a need for extended tests, or if extended tests exists, have they been implemented in an appropriate way, and are they appropriately documented?
*I think there need to be more tests, but not an explicit set of "extended tests". The additional checks for appropriate warnings/errors, while numerous, should be relatively quick to run.*

#### Visualisation (where appropriate)

- Do visualisations aid the primary purposes of statistical interpretation of results?
*N/A*

- Are there any aspects of visualisations which could risk statistical misinterpretation?
*N/A*

#### Package Design

- Is the package well designed for its intended purpose?
*Mostly. It enables the usage of a few different models for a common purpose and while there are a lot of exported functions, most users will probably only interact with a few. The main issue is the complexity of those functions.*

- In relation to **External Design:** Do exported functions and the relationships between them enable general usage of the package?
*The functions listed on the main documentation page (`stoch_crm` and `band_crm`) are extrememly complex and I think this makes them difficult to use. `stoch_crm` has 31 arguments, which actually hides the number of parameters in the model, as many of these arguments expect data frames containing multiple parameters. I recognize that this isn't the authors' fault (it's a complex model and they have worked to make it more accessible), but the resulting code is still fairly intimidating. I don't know how familiar users will be with some of these specific arguments, but if it is at all possible to provide more reasonable default values, I think that would help.*

- In relation to **External Design:** Do exported functions best serve inter-operability with other packages?
*I'm not familiar with any other packages that handle similar questions.*

- In relation to **Internal Design:** Are algorithms implemented appropriately in terms of aspects such as efficiency, flexibility, generality, and accuracy? 
*I think so.*

- In relation to **Internal Design:** Could ranges of admissible input structures, or form(s) of output structures, be expanded to enhance inter-operability with other packages?
*A little bit, yes. All of the `*_pars` arguments expect a one-row data frame with column names `"mean"` and `"sd"`, but I think these could accept numeric vectors, as well. You could require names (as in `c(mean = 0, sd = 1)`), or just make it clear in the documentation of these arguments that the first element is assumed to be the mean and the second is the SD.*

---

- [x] **Packaging guidelines**: The package conforms to the rOpenSci packaging guidelines

Estimated hours spent reviewing: 20

- [x] Should the author(s) deem it appropriate, I agree to be acknowledged as a package reviewer ("rev" role) in the package DESCRIPTION file.

---

### Review Comments

The guidelines don't seem too concerned about this, but my biggest concern with the package is the vignette. This is usually where I start when I want to use a new package, but it's very sparse and doesn't explain much. I would like for the vignette to provide more background and explanation, like the main package documentation. Also, none of the code is actually executed and some of it doesn't even run (there's a stray "messages" that should have been commented out in the Multiscenario Example. This is also present in the README).

I also want to reiterate that if there's anything the authors can do to reduce the number of arguments, provide more suggested default values, or in any way simplify the user experience, that would be greatly appreciated and make the package more approachable.

Some other minor notes:
 - In the main package documentation, there seems to be some filler text that needs to be replaced ("...fill with migration functions..")
 - There are a lot of checks that look like this:
 ```
 if(missing(bird_dens_opt)){
    bird_dens_opt <- "tnorm"
  }
 ```
 but should probably look like this:
 ```
 bird_dens_opt <- match.arg(bird_dens_opt)
 ```
and `match.arg` will automatically treat the first possible value as the default.
 - In those same checks, `stoch_crm` has two checks for `out_period`, but I'm guessing the second one was supposed to be `out_format`.
 - I don't know if this is intentional, but in the vignette and README, you provide the argument `log_file = paste0(getwd(), "scrm_example.log")`. Since this doesn't include a separator it's creating something with the name `[currentdirectory]scrm_example.log` in the parent of the current directory. If that's not the intended behavior, it might be safer to replace this with `file.path(getwd(), "scrm_example.log")`.
 - There are a few commented out functions at the end of `validate_inputs.R` that could probably be removed

**Comment by grwhumphries** [comment]
@kstreet13 - thank you very much for the extensive review of our package! All of your comments are greatly appreciated and we owe you one. 

@stephaniehicks @mpadge - what are the next steps for us?

**Comment by mpadge** [comment]
@stephaniehicks and @grwhumphries Please note that minor issues from the rOpenSci side have prevented @clairemas0n from being able to submit her review. We'll resolve our issues straight away, and review should appear here soon. Sorry for any inconvenience.

**Comment by clairemas0n** [comment]
## Package Review for 'stockLAB' by clairemas0n


Big apologies for the delay in my review. I also would like to note that this is my first time reviewing a package and it was quite a challenge. With that in mind, I don't think my review as thorough from the software/coding point of view compared to others, but perhaps my contribution is as a potential end user who would use this package to analyse my data. An emerging threat for the seabird I study is a growing number of offshore wind farm proposals popping up in its marine distribution !! 

- [X] As the reviewer I confirm that there are no [conflicts of interest](https://devguide.ropensci.org/policies.html)
                                                                            
### Compliance with Standards
                                                                            
- [X] This package complies with a sufficient number of standards for a SILVER (?) badge

                                                                          
#### The following standards currently deemed non-applicable (through tags of `@srrstatsNA`) could potentially be applied to future versions of this software: 
										
All seem fine to be non-applicable. However, I can see how G2.16, G2.4c, G2.4d, G2.4e, & G2.9 would be useful in making sure data is correct format and minimising mistakes/bugs. 
			                                                                      
#### Please also comment on any standards which you consider either particularly well, or insufficiently, documented.

G1.0 & G1.1 are very well documented. It is easy to understand what work stochLAB builds on and how. 
G5.1 datasets for tests are very useful for understanding the utility of the functions and the format for data preparation needed to use the package. 

I did not find comprehensive documentation on EA1.0, EA1.1, EA1.2, EA1.3

For packages aiming for silver or gold badges:
                                                                            
- [X] This package extends beyond minimal compliance with standards in the following ways: (please describe)
  
Documentation of functions is excellent. 
                                                                        
---
                                                                            
### General Review
                                                                            
#### Documentation
                                                                            
The package includes all the following forms of documentation:
                                                                            
- [X] **A statement of need** clearly stating problems the software is designed to solve and its target audience in README
It would be helpful to have another sentence or two here to further contextualise the package. 

- [X] **Installation instructions:** for the development version of package and any non-standard dependencies in README
No mention of non-standard dependencies - assumed to be N/A?

- [X] **Community guidelines** including contribution guidelines in the README or CONTRIBUTING
A link to report bugs at github/issues was found in ?stochLAB() in R however couldn't find this on the github.com readme version?

- [X] The documentation is sufficient to enable general use of the package beyond one specific use case
It might be helpful to have more information on how to use the outputs -e.g. example graphs or summary tables. I found it difficult to visualise/understand the stoch_crm() output. 
                                                                          
The following sections of this template include questions intended to be used as guides to provide general, descriptive responses. Please remove this, and any subsequent lines that are not relevant or necessary for your final review.
                                                                          
                                                                            
#### Package Design
                                                                            
- Is the package well designed for its intended purpose?
Yes. However for seabird biologists to use this package, I think further documentation of how to use the outputs and perhaps how to access wind farm data would be very helpful. 

- In relation to **External Design:** Do exported functions and the relationships between them enable general usage of the package? 
Yes. However, as there are many arguments in the function, some more guidance on where to obtain some of these parameters would be helpful. 

- In relation to **External Design:** Do exported functions best serve inter-operability with other packages?
Unsure. 

- In relation to **Internal Design:** Are algorithms implemented appropriately in terms of aspects such as efficiency, flexibility, generality, and accuracy? 
Yes. 

- In relation to **Internal Design:** Could ranges of admissible input structures, or form(s) of output structures, be expanded to enhance inter-operability with other packages?
Unsure. Perhaps outputs for plotting spatially in R e.g. visualising location of wind farms and movement of seabirds? 
                                                                            
---
                                                                            
- [X] **Packaging guidelines**: The package conforms to the rOpenSci packaging guidelines
                                                                          
Estimated hours spent reviewing: 5                                                                                                                                              
                                                                         
- [X] Does the package comply with the rOpenSci packaging guide?

Very exciting to see this kind of analysis as at my fingertips with this package. I can honestly see myself being able to use these models for my own analysis. 
A few things that would (personally) make this more achievable to me would be if more background was provided on the input data/arguments so I could more easily source and format data to use in the functions. And also, if more guidance or examples were provided on how to interpret and use the outputs. 

Thank you for the opportunity. Again, sorry for the delay and I hope my thoughts are helpful.

**Comment by grwhumphries** [comment]
@clairemas0n - thank you so much for your review. Your thoughts are really helpful and we'll look at these carefully. 

@stephaniehicks - what are next steps now that the reviews are in?

**Comment by stephaniehicks** [comment]
Hi @grwhumphries thank you for your patience with me! Teaching has been particularly stressful this fall, which caused the delay. 

I want to thank @kstreet13 and @clairemas0n for your time and effort reviewing this package and giving your thoughtful feedback! 

Everyone -- I will now record the reviews in new comments below.

**Comment by stephaniehicks** [comment]
@ropensci-review-bot submit review https://github.com/ropensci/software-review/issues/551#issuecomment-1340101540 time 5

**Comment by grwhumphries** [comment]
@clairemas0n - would you like to be listed in the DESCRIPTION as a reviewer?

**Comment by grwhumphries** [comment]
## Response to reviewers
Thank you so much @clairemas0n and @kstreet13 ->  your reviews have been really helpful. You've both put a significant amount of work into looking through this and we greatly appreciate it.

I have responded to specific comments as pasted here - hope that works; let me know if there are issues.

See attached response. 

### Response to @clairemas0n 

I did not find comprehensive documentation on EA1.0, EA1.1, EA1.2, EA1.3
**GH – I have added more information in the README that should hopefully contextualise these points a little better**

A statement of need clearly stating problems the software is designed to solve and its target audience in README
It would be helpful to have another sentence or two here to further contextualise the package.
**GH – I have added some more information here in the description and in the README**

Installation instructions: for the development version of package and any non-standard dependencies in README
No mention of non-standard dependencies - assumed to be N/A?
**GH – I have added a sentence in the README on the package dependencies**

Community guidelines including contribution guidelines in the README or CONTRIBUTING
A link to report bugs at github/issues was found in ?stochLAB() in R however couldn't find this on the github.com readme version?
**GH – This has now been added**

The documentation is sufficient to enable general use of the package beyond one specific use case
It might be helpful to have more information on how to use the outputs -e.g. example graphs or summary tables. I found it difficult to visualise/understand the stoch_crm() output.
**GH – I have added new sections to the README to reflect these points**

Is the package well designed for its intended purpose?
CM - Yes. However for seabird biologists to use this package, I think further documentation of how to use the outputs and perhaps how to access wind farm data would be very helpful.
**GH – I have added new sections to the README to reflect these points**

In relation to External Design: Do exported functions and the relationships between them enable general usage of the package?
CM - Yes. However, as there are many arguments in the function, some more guidance on where to obtain some of these parameters would be helpful.
**GH – I have added new sections to the README to reflect these points**

In relation to External Design: Do exported functions best serve inter-operability with other packages?
CM - Unsure.
**GH – two other packages interact with stochLAB:  mCRM (hidef-aerial-surveying/mCRM) and sCRM (dmpstats/sCRM) -> both have been designed around this package and respective outputs.**

In relation to Internal Design: Could ranges of admissible input structures, or form(s) of output structures, be expanded to enhance inter-operability with other packages?
CM - Unsure. Perhaps outputs for plotting spatially in R e.g. visualising location of wind farms and movement of seabirds?
**GH – Unfortunately, stochLAB is a-spatial and so there is no scope to add visualisation of the locations of wind farms/seabird movement.**

### Response to @kstreet13 
 
EA2.4 and EA2.5 concern the use of multitabular input, which is relevant to both stoch_crm and band_crm.
**GH – data.frame and/or tibble classes are demanded and checked, so have marked these standards as met**

G5.2, G5.2a, and G5.2b. The authors are correct that inputs are checked by the validate_inputs scripts, but this is not explicitly tested, as far as I can tell (ie. by having unit tests that give invalid inupts and checking for the correct error message).
**GH – This could be something implemented in future iterations; the validate_inputs script is very thorough and should cover most issues. Due to the number of parameters (which can’t be helped), the number of possible errors could be substantial and it would be far too time consuming for every issue to be explicitly tested; however, the validate_inputs script makes sure that all data are input in the correct format.**

G1.5 is not directly applicable, but the authors do claim in the documentation that their implementation of the extended model is "over 30x faster than Masden's code", which is not demonstrated.
**GH – Masden’s code is no longer used for collision risk modelling and we didn’t feel it a useful exercise to benchmark this. We have removed the terminology “30x faster” so that it isn’t quantified. The quantification was done by way of experience with older versions of the code**

G1.4a Internal functions are not all documented (esp. in validate_inputs)
**GH – These functions in validate_inputs are now documented**

A statement of need clearly stating problems the software is designed to solve and its target audience in README
KS - The first sentence lays out what the package does, but I'm a little unclear on who the target audience is. I'm guessing it's some combination of people who own wind farms and ornithologists?
**GH – I have added more information in the README now**

How well are algorithms encoded?
KS - The overall structure looks good to me, though some of the details (both of the method and the implementation) are lost on me. I'm guessing the big for loop that runs the model a bunch of times with different parameters is the most time-consuming part, so one potential future improvement would be to run those in parallel rather than sequentially.
**GH – we discussed parallelization of the stochastic process, but because it runs so quickly, we decided against it.** 

Testing
**GH - Our feeling on this was that the validate_inputs function does a thorough job of ensuring the functions work as expected/required. Due to the incredible number of possible ways that error messages could be generated, we felt this was the most appropriate way forward. However, future versions of the package could integrate some of these tests.** 

Is the package well designed for its intended purpose?
KS - Mostly. It enables the usage of a few different models for a common purpose and while there are a lot of exported functions, most users will probably only interact with a few. The main issue is the complexity of those functions.
**GH – Unfortunately, the complexity is a function of the number of parameters associated with the models; we couldn’t find a way around this.**

In relation to External Design: Do exported functions and the relationships between them enable general usage of the package?
KS - The functions listed on the main documentation page (stoch_crm and band_crm) are extrememly complex and I think this makes them difficult to use. stoch_crm has 31 arguments, which actually hides the number of parameters in the model, as many of these arguments expect data frames containing multiple parameters. I recognize that this isn't the authors' fault (it's a complex model and they have worked to make it more accessible), but the resulting code is still fairly intimidating. I don't know how familiar users will be with some of these specific arguments, but if it is at all possible to provide more reasonable default values, I think that would help.
**GH – agreed that the function is complicated. The package itself is wrapped up inside of two other R Shiny applications, and we have added some more information in the README to help users along with where to get data.** 

In relation to Internal Design: Could ranges of admissible input structures, or form(s) of output structures, be expanded to enhance inter-operability with other packages?
KS - A little bit, yes. All of the *_pars arguments expect a one-row data frame with column names "mean" and "sd", but I think these could accept numeric vectors, as well. You could require names (as in c(mean = 0, sd = 1)), or just make it clear in the documentation of these arguments that the first element is assumed to be the mean and the second is the SD.
**GH – Although practically, yes, this could be rearranged so that it accepts the vector argument, the val_pars_df() function which validates this tests if this is a data frame or tibble, and how many rows are in that data frame. It also tests that column names are exactly “mean” or “sd” to ensure consistency.**

The guidelines don't seem too concerned about this, but my biggest concern with the package is the vignette. This is usually where I start when I want to use a new package, but it's very sparse and doesn't explain much. I would like for the vignette to provide more background and explanation, like the main package documentation. Also, none of the code is actually executed and some of it doesn't even run (there's a stray "messages" that should have been commented out in the Multiscenario Example. This is also present in the README).
**GH – The README examples are now executed**

I also want to reiterate that if there's anything the authors can do to reduce the number of arguments, provide more suggested default values, or in any way simplify the user experience, that would be greatly appreciated and make the package more approachable.
**GH – I have added more information in the README now that should hopefully help to clear up some of these issues and make it more accessible. Unfortunately there isn’t a way to limit the number of parameters due to the nature of the model**

In the main package documentation, there seems to be some filler text that needs to be replaced ("...fill with migration functions..")
**GH – These filler texts have been removed**

match.arg functions
**GH – Thanks for catching that, I have integrated that now**

In those same checks, stoch_crm has two checks for out_period, but I'm guessing the second one was supposed to be out_format.
**GH – Thanks for catching that, now changed**

I don't know if this is intentional, but in the vignette and README, you provide the argument log_file = paste0(getwd(), "scrm_example.log"). Since this doesn't include a separator it's creating something with the name [currentdirectory]scrm_example.log in the parent of the current directory. If that's not the intended behavior, it might be safer to replace this with file.path(getwd(), "scrm_example.log").
**GH – Thanks, have updated this in the vignette and README**

There are a few commented out functions at the end of validate_inputs.R that could probably be removed
**GH – Thanks, have removed these**

### Points for clarification
PD3.5 seems to concern integration generally, which is a part of the model.
**GH – Could use some clarification here as I’m not sure how this standard is applicable to the model exactly.** 

G1.3 Details of the models including statistical terminology are not always explained, but users are directed to appropriate sources, such as Masden (2015) and Band (2012).
**GH – The main users of the methodology should be familiar with these papers as they are critical for understanding how results need to be discussed for environmental impact assessments. Were there specific terms that needed clarification?**

**Comment by stephaniehicks** [comment]
Thanks @grwhumphries! 

@kstreet13 @clairemas0n Can you review @grwhumphries's response and let us know if the updates resolve any outstanding issues for acceptance at gold level? Thanks!

**Comment by clairemas0n** [comment]
Thank you @grwhumphries for taking the time to consider my comments. The README is now very comprehensive and contextualises the package very well. I acknowledge my review has been responded to and I approve acceptance at a gold level. Congratulations @grwhumphries on all your work getting this package to this standard. Very impressive! 

> @clairemas0n - would you like to be listed in the DESCRIPTION as a reviewer?
Sorry missed this. I wasn't sure whether my contributions were substantial enough to be considered a reviewer. But happy to be mentioned if they contributed enough to the development of your package. Not worried either way. Thank you @grwhumphries. 

 @ropensci-review-bot submit review <https://github.com/ropensci/software-review/issues/551#issuecomment-1340101540> time 5

**Comment by grwhumphries** [comment]
HI @stephaniehicks @kstreet13 
just wanted to check to see if there's been any movement on the final checks for the package? Would be keen to tie this up at some point soon :) 

Cheers and thanks for your efforts on this

**Comment by maurolepore** [comment]
Dear all,

I'm the EiC on duty and now also acting as handling editor -- thanks @stephaniehicks for your work!

@kstreet13 could you please let me know if you're happy with the author's response and either request further changes or indicate approval with the [reviewer approval template](https://devguide.ropensci.org/approval2template.html)?

**Comment by kstreet13** [comment]
Sorry! I somehow missed this in December. I think the package is much improved and all of my concerns have been addressed.

Regarding the points for clarification:
 - for PD3.5, I think I was looking at `get_prop_crh_fhd`, which contains the comment `# integration over the rotor's height bands`, but I think the height bands are discrete, so this isn't actually an approximation of a continuous integral.
 - Model terms are sufficiently explained, I think I was just confused by the complexity and not sufficiently familiar with the original papers.

I am happy to accept this package at a *gold* level. I don't think the standards for testing have been met in the way that they are written, but I absolutely agree with @grwhumphries that to do so would be impractical for this package and the current implementation is more than sufficient.

## Reviewer Response

#### Final approval (post-review)

- [x] **The author has responded to my review and made changes to my satisfaction. I recommend approving this package.**

Estimated hours spent reviewing: 21

**Comment by maurolepore** [comment]
@ropensci-review-bot approve stochLAB

**Comment by maurolepore** [comment]
@grwhumphries, 

Congratulations! 

I believe you won't be able to check To-do's above so here they are again in plain text. You can copy-paste them in a comment of your own and check-chem as you complete them. 

Let me know if you're blocked with any.

```
- [ ] Transfer the repo to rOpenSci's "ropensci" GitHub organization under "Settings" in your repo.  I have invited you to a team that should allow you to do so. You will need to [enable two-factor authentication](https://docs.github.com/en/authentication/securing-your-account-with-two-factor-authentication-2fa/configuring-two-factor-authentication) for your GitHub account.
**This invitation will expire after one week. If it happens write a comment `@ropensci-review-bot invite me to ropensci/<package-name>` which will re-send an invitation.**
- [ ] After transfer write a comment `@ropensci-review-bot finalize transfer of <package-name>` where `<package-name>` is the repo/package name. This will give you admin access back.
- [ ] Fix all links to the GitHub repo to point to the repo under the ropensci organization.
- [ ] Delete your current code of conduct file if you had one since rOpenSci's default one will apply, see https://devguide.ropensci.org/collaboration.html#coc-file
- [ ] If you already had a `pkgdown` website **and are ok relying only on [rOpenSci central docs building and branding](https://devguide.ropensci.org/ci.html#even-more-ci-ropensci-docs)**,
    * deactivate the automatic deployment you might have set up
    * remove styling tweaks from your pkgdown config but keep that config file
    * replace the whole current `pkgdown` website with a [redirecting page](https://devguide.ropensci.org/redirect.html)
    * replace your package docs URL with `https://docs.ropensci.org/package_name`
    * In addition, in your DESCRIPTION file, include the docs link in the `URL` field alongside the link to the GitHub repository, e.g.: `URL: https://docs.ropensci.org/foobar, https://github.com/ropensci/foobar`
- [ ] Skim the docs of the [pkgdown automatic deployment](https://devguide.ropensci.org/building.html#docsropensci), in particular if your website needs MathJax.
- [ ] Fix any links in badges for CI and coverage to point to the new repository URL. 
- [ ] Increment the package version to reflect the changes you made during review. In NEWS.md, add a heading for the new version and one bullet for each user-facing change, and each developer-facing change that you think is relevant.
- [ ] We're starting to roll out software metadata files to all rOpenSci packages via the Codemeta initiative, see https://docs.ropensci.org/codemetar/ for how to include it in your package, after installing the package - should be easy as running `codemetar::write_codemeta()` in the root of your package.
- [ ] You can add this installation method to your package README `install.packages("<package-name>", repos = "https://ropensci.r-universe.dev")` thanks to [R-universe](https://ropensci.org/blog/2021/06/22/setup-runiverse/).
```

**Comment by maelle** [comment]
@grwhumphries did you meet any difficulty? I'm asking as I'm drafting rOpenSci newsletter I'll send on Friday: if your package is transferred by then, it will be mentioned in this month's newsletter (if not, in next month's so no worries).

**Comment by grwhumphries** [comment]
@ropensci-review-bot finalize transfer of grwhumphries/stochLAB

**Comment by grwhumphries** [comment]
@ropensci-review-bot finalize transfer of  ropensci/stochLAB

**Comment by grwhumphries** [comment]
@ropensci-review-bot finalize transfer of ropensci/stochLAB

**Comment by grwhumphries** [comment]
@maurolepore @maelle @stephaniehicks any idea why the owner rights aren't transferring properly?

**Comment by maelle** [comment]
@grwhumphries you first need to transfer the repository to rOpenSci via the repository settings (I've just checked that you had accepted the invitation to the ropensci GitHub organization so you have the rights to do that). 

This will transfer the repo but lock you out. To get admin access back, you will then write a comment like https://github.com/ropensci/software-review/issues/551#issuecomment-1441504788

Please tell me how it goes + if anything is unclear.

**Comment by grwhumphries** [comment]
Hi @maelle - I've already transferred the repo over to rOpenSci  (check out ropensci/stochLAB).  Then I wrote `@ropensci-review-bot finalize transfer of ropensci/stochLAB`  - but still getting: `Could not finalize transfer: Could not add owner rights to the ropensci/stochLAB team`

**Comment by grwhumphries** [comment]
@ropensci-review-bot finalize transfer of stochLAB

**Comment by grwhumphries** [comment]
AH! This needs to be in the format:  `@ropensci-review-bot finalize transfer of stochLAB` ->  the instructions suggested this had to be `ropensci/stochLAB`

**Comment by maelle** [comment]
@grwhumphries this is due to a confusion with the "invite me" command, correct? or which instructions do you mean? (I agree the two different syntaxes are confusing)
